### PR TITLE
Add support for ISet<T>

### PR DESCRIPTION
**Pull request by layomia**
The approach is to treat it like HashSet<T>, similar to how we treat `IDictionary<string, TValue>` and `IReadOnlyDictionary<string, TValue>` as `Dictionary<string, TValue>`.

#### Perf
We might see some perf gain by writing a special-case converter for `ISet<T>`. ~~This will also help mitigate the limitation on (de)serializing `ISet<ISet<T>>` (and higher degrees of `ISet<T>` nesting) that I [describe in this PR](https://github.com/layomia/corefx/blob/8bb1e1b4ab55a663bd7447d0350dfd13813718b9/src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs#L98-L104).~~
For now, I'd rather spend the time on getting this and other collections through (https://github.com/dotnet/corefx/issues/36643), and circle back during/following testing/validation.

##### `ISet<T>` (de)serialization

The benchmarks are being added to the perf repo: https://github.com/dotnet/performance/pull/540.

|                Method | ElementCount |        Mean |      Error |     StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |------------- |------------:|-----------:|-----------:|-------:|------:|------:|----------:|
|       **DeserializeISet** |            **2** |  **2,114.5 ns** |  **41.533 ns** |  **44.440 ns** | **0.2747** |     **-** |     **-** |    **1160 B** |
| SerializeISet_ToBytes |            2 |    507.2 ns |   5.404 ns |   4.791 ns | 0.0591 |     - |     - |     248 B |
|       **DeserializeISet** |           **50** | **11,712.8 ns** | **252.800 ns** | **346.035 ns** | **1.8616** |     **-** |     **-** |    **7817 B** |
| SerializeISet_ToBytes |           50 |  5,993.0 ns | 106.807 ns |  99.907 ns | 0.2823 |     - |     - |    1192 B |
|       **DeserializeISet** |          **100** | **21,081.3 ns** | **124.253 ns** | **103.757 ns** | **3.6621** |     **-** |     **-** |   **15353 B** |
| SerializeISet_ToBytes |          100 | 11,535.3 ns | 173.335 ns | 135.328 ns | 0.5188 |     - |     - |    2192 B |

##### Json.Net

|                Method | ElementCount |        Mean |       Error |       StdDev |      Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |------------- |------------:|------------:|-------------:|------------:|-------:|------:|------:|----------:|
|       **DeserializeISet** |            **2** | **21,030.9 ns** | **3,649.67 ns** | **10,412.72 ns** | **17,650.0 ns** |      **-** |     **-** |     **-** |   **2.84 KB** |
| SerializeISet_ToBytes |            2 |    542.6 ns |    10.85 ns |     23.35 ns |    536.1 ns | 0.3328 |     - |     - |   1.36 KB |
|       **DeserializeISet** |           **50** |  **9,447.0 ns** |   **188.40 ns** |    **224.27 ns** |  **9,434.3 ns** | **1.9073** |     **-** |     **-** |   **7.85 KB** |
| SerializeISet_ToBytes |           50 |  4,475.9 ns |    86.77 ns |     72.46 ns |  4,493.8 ns | 0.6943 |     - |     - |   2.85 KB |
|       **DeserializeISet** |          **100** | **18,215.9 ns** |    **70.56 ns** |     **62.55 ns** | **18,209.3 ns** | **3.3264** |     **-** |     **-** |   **13.7 KB** |
| SerializeISet_ToBytes |          100 |  8,535.8 ns |    70.04 ns |     65.51 ns |  8,553.2 ns | 1.1902 |     - |     - |    4.9 KB |

```C#
public class JsonDotNet_ReadGenericCollection
{
    private static readonly ISet<string> _iset = new HashSet<string>();
    private static string _jsonString;

    [Params(2, 50, 100)]
    public int ElementCount;

    [GlobalSetup]
    public void Setup()
    {
        for (int i = 0; i < ElementCount; i++)
        {
            _iset.Add($"hello{i}");
        }

        _jsonString = System.Text.Json.Serialization.JsonSerializer.ToString(_iset);
    }

    [Benchmark]
    public ISet<string> DeserializeISet()
    {
        return JsonConvert.DeserializeObject<ISet<string>>(_jsonString);
    }

    [Benchmark]
    public string SerializeISet_ToBytes()
    {
        return JsonConvert.SerializeObject(_iset);
    }
}
```

##### Hot-path (de)serialization scenarios 
Benchmarks live in the [perf repo](https://github.com/dotnet/performance/tree/master/src/benchmarks/micro/corefx/System.Text.Json/Serializer).
No regressions: [hot-path benchmark results.zip](https://github.com/dotnet/corefx/files/3253143/hot-path.benchmark.results.zip)

**UPDATE**
The limitation on (de)serializing `ISet<ISet<T>>` is fixed by using a runtime property type of `HashSet<TDeclaredPropertyOfElement>` instead of `HashSet<TRuntimePropertyOfElement>` for `ISet<TDeclaredPropertyOfElement>`

### CODE REVIEW

**Comment by steveharter** [comment]
On the benchmarks - Json.NET takes longer to deserialize 2 items than 50 or 100?

**Comment by layomia** [comment]
> On the benchmarks - Json.NET takes longer to deserialize 2 items than 50 or 100?

Yes, even on repeated runs. I’m looking into why.

File: src/System.Text.Json/tests/Serialization/Value.WriteTests.GenericCollections.cs [hunk 1]
```diff
@@ -285,6 +285,77 @@ public static void WritePrimitiveIReadOnlyListT()
...
+        public static void WriteISetTOfHashSetT()
+        {
+            ISet<HashSet<int>> input = new HashSet<HashSet<int>>
+            {
+                new HashSet<int>() { 1, 2 },
+                new HashSet<int>() { 3, 4 }
+            };
+
+            string json = JsonSerializer.ToString(input);
+            Assert.Equal("[[1,2],[3,4]]", json);
```
Comment (watfordgnf) [inline src/System.Text.Json/tests/Serialization/Value.WriteTests.GenericCollections.cs]: `"[[3,4],[1,2]]"` is also valid (as is `"[[4,3],[1,2]]"`, etc); is there a way to structure these tests without the order dependence? Or is that okay given HashSet's implementation details are likely "fixed" (presuming folks depend on insert order being enumeration order).

File: src/System.Text.Json/tests/Serialization/PolymorphicTests.cs [hunk 2]
```diff
@@ -130,6 +130,13 @@ public static void ArrayAsRootObject()
             json = JsonSerializer.ToString<object>(ireadonlylist);
             Assert.Equal(ExpectedJson, json);
 
+            ISet<object> iset = new HashSet<object> { 1, true, address, null, "foo" };
+            json = JsonSerializer.ToString(iset);
+            Assert.Equal(ExpectedJson, json);
```
Comment (watfordgnf) [inline src/System.Text.Json/tests/Serialization/PolymorphicTests.cs]: Order isn't guaranteed, right? Looks like the tests for `ImmutableSet` which use `expectedObjects` are a better way to test.

(Same for HashSet tests in this file, which I realize you did not add.)

File: src/System.Text.Json/tests/Serialization/Value.ReadTests.GenericCollections.cs [hunk 3]
```diff
@@ -354,6 +354,87 @@ public static void ReadPrimitiveIReadOnlyListT()
...
+
+        [Fact]
+        public static void ReadISetTOfHashSetT()
+        {
+            ISet<HashSet<int>> result = JsonSerializer.Parse<ISet<HashSet<int>>>(Encoding.UTF8.GetBytes(@"[[1,2],[3,4]]"));
+            int expected = 1;
+
+            foreach (HashSet<int> ie in result)
+            {
+                foreach (int i in ie)
```
[the remark this answers is not in the document]
Comment (Clockwork-Muse) [reply src/System.Text.Json/tests/Serialization/Value.ReadTests.GenericCollections.cs]: Note that `Xunit.Assert.Equal(...)` _does_ handle set equality as matching unordered elements.  Unfortunately it's not recursive (it checks based on element equality).  Still, you can make it the slightly less brittle

```csharp
Assert.Equal(result.First(), new HashSet<int> {2, 1});
Assert.Equal(result.Last(), new HashSet<int> {4, 3});
```

**Comment by ahsonkhan** [comment]
> On the benchmarks - Json.NET takes longer to deserialize 2 items than 50 or 100?

> Yes, even on repeated runs. I’m looking into why. First guess is cached reflection.

Looking at the Error/StdDev, that test looks noisy. You want higher warmup count probably.

File: src/System.Text.Json/tests/Serialization/Value.WriteTests.GenericCollections.cs (hunk 1 above)
Comment (layomia) [reply src/System.Text.Json/tests/Serialization/Value.WriteTests.GenericCollections.cs]: I've structured these tests without the order dependence - I think it's best.

**Comment by layomia** [comment]
Changes here have been merged in https://github.com/dotnet/corefx/pull/38319.

**Pull request closed without merging**